docs/sdk2/usage/PRODUCT_SYNC.md

```diff
@@ -0,0 +1,471 @@
+CompletableFuture<List<ProductDraft>> productDrafts = ProductTransformUtils.toProductDrafts(client, referenceIdToKeyCache, products);
+````
+##### Syncing from an external resource
+
+- When syncing from an external resource, `ResourceIdentifier`s with their `key`s have to be supplied as following example:
+
+```` java
+final ProductDraft productDraft =
+              ProductDraftBuilder
+                      .of()
+                      .productType(ProductTypeResourceIdentifierBuilder.of().key("product-type-key").build())
+                      .name(LocalizedString.ofEnglish("name"))
+                      .slug(LocalizedString.ofEnglish("slug"))
+                      .masterVariant(masterVariant)
+                      .key("product-key")
+                      .categories(CategoryResourceIdentifierBuilder.of().key("category1-key").build(),
+                              CategoryResourceIdentifierBuilder.of().key("category2-key").build())
+                      .taxCategory(TaxCategoryResourceIdentifierBuilder.of().key("tax-category-key").build())
+                      .state(StateResourceIdentifierBuilder.of().key("tax-category-key").build())
+                      .build();
+````
+ 
+ ````java
+final PriceDraft priceDraft =
+          PriceDraftBuilder.of()
+              .value(MoneyBuilder.of().centAmount(20L).currencyCode("EUR").build())
+              .channel(ChannelResourceIdentifierBuilder.of().key("channel-key").build())
+              .customerGroup(
+                  CustomerGroupResourceIdentifierBuilder.of().key("customer-group-key").build())
+              .custom(
+                  CustomFieldsDraftBuilder.of()
+                      .type(TypeResourceIdentifierBuilder.of().key("type-key").build())
+                      .fields(FieldContainerBuilder.of().values(Collections.emptyMap()).build()).build())
+              .build();
+````
+
+-  The product projection variant attributes with a type `ReferenceType` do not support the `ResourceIdentifier` yet, for
+ those references you have to provide the `key`  value on the `id` field of the reference. This means that calling `getId()` on the reference should return its `key`.
+
+````java
+final ProductReference productReference = ProductReferenceBuilder.of().id("product-key").build();
+final Attribute attr = AttributeBuilder.of().name("attribute-name").value(productReference).build();
+
+````
+
+-  For resolving `key-value-document` (custom object) references on attributes of type `Reference`, `Set` of `Reference`, `NestedType` or `Set` of `NestedType`, The `id` field of the reference in the attribute should be defined in the correct format. 
+The correct format must have a vertical bar `|` character between the values of the container and key.
+For example, if the custom object has a container value `container` and key value `key`, the `id` field should be `container|key"`,  
+also, the key and container value should match the pattern `[-_~.a-zA-Z0-9]+`.
```

Comment:
There is also max length for the key for custom objects (256 chars). Maybe it also matters here? https://docs.commercetools.com/api/projects/custom-objects#customobject

Reply:
Yeah, in indirect way it does matter... i will add a note.